cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java

```diff
@@ -286,4 +296,48 @@ public void processMessage(MetadataMessage message) {
       }
     }
   }
+
+  private class MetadataOperationProcessor implements MetadataMessageProcessor {
+
+    @Override
+    public void processMessage(MetadataMessage message)  {
+      MetadataOperation operation = message.getPayload(GSON, MetadataOperation.class);
+      Metadata metadata = operation.getMetadata();
+      MetadataEntity entity = operation.getEntity();
+
+      // TODO: Authorize that the operation is allowed. Currently MetadataMessage does not carry user info
+      switch (operation.getType()) {
+        case PUT: {
+          LOG.trace("Received PUT for entity {}: {}", entity, metadata);
+          try {
+            if (metadata.getProperties() != null && !metadata.getProperties().isEmpty()) {
+              metadataAdmin.addProperties(entity, metadata.getProperties());
+            }
+            if (metadata.getTags() != null && !metadata.getTags().isEmpty()) {
+              Set<String> toAdd = metadata.getTags();
+              metadataAdmin.addTags(entity, toAdd);
+
+            }
+          } catch (InvalidMetadataException e) {
+            LOG.warn("Received invalid metadata operation from TMS: {}", GSON.toJson(message.getRawPayload()), e);
```

Comment:
Should we say "Ignoring invalid metadata ..." to make it clear the payload has been skipped?

Reply:
Sure.